File: src/etch/bindings/java/msg/test/TestArrayValue.java
/* $Id$
 *
 * Created by sccomer on Feb 4, 2007.
 *
 * Copyright (c) 2007 Cisco Systems, Inc. All rights reserved.
 */

package etch.bindings.java.msg.test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

import metreos.util.FlexBuffer;

import org.junit.Test;

import etch.bindings.java.msg.ArrayValue;
import etch.bindings.java.msg.Field;
import etch.bindings.java.msg.Message;
import etch.bindings.java.msg.StructValue;
import etch.bindings.java.msg.TaggedDataOutput;

/**
 * Tests ArrayValue.
 */
public class TestArrayValue
{
	/** */
	@Test
	public void addMany()
	{
		ArrayValue av = new ArrayValue().addMany( 11, 22, 33 );
		assertEquals( 3, av.size() );
		assertEquals( 11, av.get( 0 ) );
		assertEquals( 22, av.get( 1 ) );
		assertEquals( 33, av.get( 2 ) );
	}

//	/** @throws IOException */
//	@Test
//	public void read() throws IOException
//	{
//		readHelper();
//		readHelper( 1 );
//		readHelper( 2, 3.3 );
//		readHelper( 4, true, 6 );
//		readHelper( "a", "b", "c", "d" );
//	}

	/** @throws IOException */
	@Test
	public void writeArray() throws IOException
	{
		writeHelper();
		writeHelper( 1 );
		writeHelper( 2, 3.3 );
		writeHelper( 4, true, 6 );
		writeHelper( "a", "b", "c", "d" );
	}
	
//	private void readHelper( Object... vals ) throws IOException
//	{
//		FakeTdi tdi = new FakeTdi( vals );
//		compareLists( tdi.readArrayValue(), vals );
//		tdi.close();
//	}
	
	private void writeHelper( Object... vals ) throws IOException
	{
		ArrayValue av = new ArrayValue();
		for (Object o: vals)
			av.add( o );
		FakeTdo tdo = new FakeTdo( vals );
		av.write( tdo );
		tdo.close();
	}

	/**
	 * Compare a list for having known contents.
	 * @param list
	 * @param objects
	 */
	static void compareLists( List<Object> list, Object... objects )
	{
		assertEquals( objects.length, list.size() );
		Iterator<Object> j = list.iterator();
		for (Object o: objects)
			assertEquals( o, j.next() );
		assertFalse( j.hasNext() );
	}
	
	/**
	 * TaggedDataOutput for testing.
	 */
	public static class FakeTdo implements TaggedDataOutput
	{
		/**
		 * @param vals
		 */
		public FakeTdo( Object[] vals )
		{
			this.vals = vals;
		}
		
		private final Object[] vals;

		public void startArray( ArrayValue array ) throws IOException
		{
			assertFalse( started );
			assertFalse( ended );
			assertNotNull( array );
			
			started = true;
			list = new ArrayList<Object>();
			xarray = array;
		}
		
		private boolean started;
		
		private boolean ended;
		
		private List<Object> list;
		
		private ArrayValue xarray;

		public void writeArrayElement( Object value ) throws IOException
		{
			assertTrue( started );
			assertFalse( ended );
			
			list.add( value );
		}

		public void endArray( ArrayValue array ) throws IOException
		{
			assertTrue( started );
			assertFalse( ended );
			assertEquals( xarray, array );
			
			ended = true;
		}

		/**
		 * @throws IOException
		 */
		public void close() throws IOException
		{
			assertTrue( started );
			assertTrue( ended );
			compareLists( list, vals );
		}

		public void endMessage( Message msg ) throws IOException
		{
			throw new UnsupportedOperationException();
		}

		public void endStruct( StructValue struct ) throws IOException
		{
			throw new UnsupportedOperationException();
		}

		public void startMessage( Message msg ) throws IOException
		{
			throw new UnsupportedOperationException();
		}

		public void startStruct( StructValue struct ) throws IOException
		{
			throw new UnsupportedOperationException();
		}

		public void writeStructElement( Field key, Object value ) throws IOException
		{
			throw new UnsupportedOperationException();
		}

		public void setBuffer( FlexBuffer msgBuf )
		{
			// TODO Auto-generated method stub
			
		}
	}
}

File: src/etch/bindings/csharp/msg/test/TestArrayValue.cs
///$Id$
///
///Created by Champakesan, Badri Narayanan on Jun 13, 2007.
///
///Copyright (c) 2007 Cisco Systems, Inc. All rights reserved.


using NUnit.Framework;
using System;
using System.Collections.Generic;
using Etch.Transport;

namespace Etch.Msg.Test
{

    [TestFixture]
    public class TestArrayValue 
    {

        [Test]
	    public void Constructor1()
	    {
		    ArrayValue av = new ArrayValue( new sbyte[0] );
		    Assert.AreEqual( (sbyte) 0, av.Type );
		    Assert.IsNull( av.CustomStructType );
		    Assert.AreEqual( 0, av.Dim );
	    }
    	
	    [Test]
	    public void Constructor2()
	    {
		    XType t1 = new XType( "t1" );
		    ArrayValue av = new ArrayValue( new sbyte[0], (sbyte) 1, t1, 2 );
		    Assert.AreEqual( (sbyte) 1, av.Type );
		    Assert.AreEqual( t1, av.CustomStructType );
		    Assert.AreEqual( 2, av.Dim );
	    }
    	
	    [Test]
	    public void constructor3()
	    {
		    XType t2 = new XType( "t2" );
            ArrayValue av = new ArrayValue( new sbyte[ 0 ], ( sbyte ) 3, t2, 4 );
		    Assert.AreEqual( (byte) 3, av.Type );
		    Assert.AreEqual( t2, av.CustomStructType );
		    Assert.AreEqual( 4, av.Dim );
	    }
    	
	    [Test]
	    public void AddMany1()
	    {
		    DateTime dt = new DateTime();
		    ArrayValue av = new ArrayValue( new Object[] { false, (sbyte) 1, (short) 2, 3, (long) 4, (float) 1.5, 1.6, "abc", null, dt } );
		    Assert.AreEqual( 10, av.Size() );
		    int index = 0;
		    Assert.AreEqual( false, av.Get( index++ ) );
		    Assert.AreEqual( (sbyte) 1, av.Get( index++ ) );
		    Assert.AreEqual( (short) 2, av.Get( index++ ) );
		    Assert.AreEqual( 3, av.Get( index++ ) );
		    Assert.AreEqual( (long) 4, av.Get( index++ ) );
		    Assert.AreEqual( (float) 1.5, av.Get( index++ ) );
		    Assert.AreEqual( 1.6, av.Get( index++ ) );
		    Assert.AreEqual( "abc", av.Get( index++ ) );
		    Assert.AreEqual( null, av.Get( index++ ) );
		    Assert.AreEqual( dt, av.Get( index++ ) );
	    }
    	
	    [Test]
	    public void AddMany2()
	    {
		    DateTime dt = new DateTime();
		    ArrayValue av = new ArrayValue( new Object[] { (float) 1.5, 1.6, "abc", null, dt, false, (sbyte) 1, (short) 2, 3, (long) 4 } );
		    Assert.AreEqual( 10, av.Size() );
		    int index = 0;
		    Assert.AreEqual( (float) 1.5, av.Get( index++ ) );
		    Assert.AreEqual( 1.6, av.Get( index++ ) );
		    Assert.AreEqual( "abc", av.Get( index++ ) );
		    Assert.AreEqual( null, av.Get( index++ ) );
		    Assert.AreEqual( dt, av.Get( index++ ) );
		    Assert.AreEqual( false, av.Get( index++ ) );
		    Assert.AreEqual( (sbyte) 1, av.Get( index++ ) );
		    Assert.AreEqual( (short) 2, av.Get( index++ ) );
		    Assert.AreEqual( 3, av.Get( index++ ) );
		    Assert.AreEqual( (long) 4, av.Get( index++ ) );
	    }
    }
}